### PR TITLE
Support custom configuration

### PR DESCRIPTION
**Pull request by arturo-seijas**
Applicable spec: <link>

### Overview

Support specifying a repository through charm configuration containing the configuration files to be added or overwriten

### Rationale

Configuration customization is needed for normal operation of Wazuh

### Juju Events Changes

* Added config changed handler

### Module Changes

charm.py, state.py and wazuh.py

### Library Changes

N/A

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [ ] The documentation for charmhub is updated
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -85,7 +90,12 @@ def reconcile(self) -> None:
             self.certificates.private_key,
             self.state.certificate,
         )
+        if self.state.git_repository:
+            wazuh.configure_git(container, str(self.state.git_repository), self.state.git_ssh_key)
```
Comment (jdkandersson) [inline src/charm.py:94]: What is the reason for needing to convert to `str` here please?

File: src/charm.py [hunk 2]
```diff
@@ -85,7 +90,12 @@ def reconcile(self) -> None:
...
+            wazuh.pull_configuration_files(container)
         wazuh.update_configuration(container, self.state.indexer_ips)
+        if self.state.git_repository:
+            wazuh.configure_git(container, str(self.state.git_repository), self.state.git_ssh_key)
```
Comment (jdkandersson) [inline src/charm.py:98]: This looks like a duplicate from above?

File: src/state.py [hunk 3]
```diff
@@ -67,8 +143,24 @@ def from_charm(
                 else "[]"
             )
             certificates = json.loads(certificates_json)
+            args = {key.replace("-", "_"): value for key, value in charm.config.items()}
+            valid_config = WazuhConfig(**args)  # type: ignore
```
Comment (jdkandersson) [inline src/state.py:147]: What is the reason for `type: ignore` here please?

File: tests/integration/conftest.py
Comment (jdkandersson) [inline tests/integration/conftest.py:1]: Do you think we should have an integration test with the new configurations?

File: src/state.py [hunk 4]
```diff
@@ -26,16 +27,91 @@ class InvalidStateError(Exception):
     """Exception raised when a charm configuration is found to be invalid."""
 
 
+class ProxyConfig(BaseModel):  # pylint: disable=too-few-public-methods
+    """Configuration for accessing Synapse through proxy.
```
Comment (mthaddon) [inline src/state.py:31]: Synapse?

File: src/wazuh.py [hunk 5]
```diff
@@ -73,3 +79,51 @@ def install_certificates(container: ops.Container, public_key: str, private_key:
...
+def configure_git(container: ops.Container, git_repository: str, git_ssh_key: str) -> None:
+    """Configure git.
+
+    Args:
+        container: the container to configure git for.
+        git_repository: the git repository to add to known hosts.
+        git_ssh_key: the SSH key for the git repository.
+    """
+    hostname = git_repository.split("@")[1].split("/")[0]
+    process = container.exec(["ssh-keyscan", "-t", "rsa", hostname])
```
Comment (mthaddon) [inline src/wazuh.py:93]: Does this mean we're depending on the type of ssh key being "rsa"? If so are we documenting that or checking that somewhere?

File: tests/integration/conftest.py
Comment (arturo-seijas) [reply tests/integration/conftest.py:1]: We should, but it is really slow to test this locally due to the required kernel options for opensearch, so I'll first test the functionality in PS6

File: charmcraft.yaml [hunk 6]
```diff
@@ -92,4 +90,18 @@ requires:
   ingress:
     interface: traefik_route
     limit: 1
-
+config:
+  options:
+    git-repository:
+      type: string
+      description: >
+        The repository URL where the DNS records are stored. The username has to be provided as in
```
[the remark this answers is not in the document]
Comment (arturo-seijas) [reply charmcraft.yaml:98]: Updated

File: src/charm.py (hunk 2 above)
Comment (arturo-seijas) [reply src/charm.py:98]: Fixed

File: src/state.py (hunk 3 above)
Comment (arturo-seijas) [reply src/state.py:147]: Added a comment with the explanation

File: src/charm.py (hunk 1 above)
Comment (arturo-seijas) [reply src/charm.py]: Actually, I don't need it. Removed

File: src/state.py (hunk 4 above)
Comment (arturo-seijas) [reply src/state.py:31]: Oops

File: src/wazuh.py (hunk 5 above)
Comment (arturo-seijas) [reply src/wazuh.py:93]: We need to add the hostname to the .known_hosts file and that's the way :)

Comment (mthaddon) [reply src/wazuh.py:93]: You're running `ssh-keyscan` with "type == rsa", so isn't that going to fail if someone specifies a key of a different type (e.g. `ed25519`)?

Comment (arturo-seijas) [reply src/wazuh.py:93]: This doesn't validate the user's key but the server's key. It's equivalent to confirming "yes" when opening an SSH connection for the first time

Comment (mthaddon) [reply src/wazuh.py:93]: I see, thx.

File: charmcraft.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (arturo-seijas) [reply charmcraft.yaml]: Renamed

File: charmcraft.yaml [hunk 7]
```diff
@@ -92,4 +90,20 @@ requires:
...
     limit: 1
-
+config:
+  options:
+    custom-config-repository:
+      type: string
+      description: >
+        The repository URL where the Wazuh configuration is stored. The username has to be provided as in 
+        git+ssh://username@repository@branch, where the branch is optional.
+        The charms expects the repository to contain a /var/ossec directory to be copied over /var/ossec on
```
Comment (mthaddon) [inline charmcraft.yaml:100]: I'd suggest dropping the first "/". "The charms expects the repository to contain a var/ossec directory to be copied over /var/ossec"

**Review by mthaddon — DISMISSED** [review]
LGTM, thx

File: charmcraft.yaml [hunk 8]
```diff
@@ -92,4 +90,20 @@ requires:
...
     interface: traefik_route
     limit: 1
-
+config:
+  options:
+    custom-config-repository:
+      type: string
+      description: >
+        The repository URL where the Wazuh configuration is stored. The username has to be provided as in 
+        git+ssh://username@repository@branch, where the branch is optional.
```
Comment (yanksyoon) [inline charmcraft.yaml:99]: Out of curiosity, do all configuration repositories have to be public?

File: charmcraft.yaml [hunk 9]
```diff
@@ -92,4 +90,20 @@ requires:
...
+        The charms expects the repository to contain a var/ossec directory to be copied over /var/ossec on
+        the running instance.
+    custom-config-ssh-key:
+      type: secret
+      description: >
+        The Juju secret ID corresponding to the private key for SSH authentication.
+        The secret should contain a single key, "value", which maps to the actual SSH key. 
+        To create the secret, run the following command: 
+        `juju add-secret my-custom-config-ssh-key value=<ssh-key> && juju grant-secret my-custom-config-ssh-key wazuh-server`,
+        and use the outputted secret ID to configure this option.
```
Comment (yanksyoon) [inline charmcraft.yaml:109]: s/the outputted secret ID/the secret ID output/

File: tests/unit/test_charm.py [hunk 10]
```diff
@@ -29,17 +29,33 @@ def test_invalid_state_reaches_blocked_status(state_from_charm_mock):
...
+    state_from_charm_mock,
 ):
     """
     arrange: mock system calls and charm state.
     act: call reconcile.
     assert: the charm reaches active status and configs are applied.
     """
-    state_from_charm_mock.return_value = State(certificate="somecert", indexer_ips=["10.0.0.1"])
+    custom_config_repository = "git+ssh://user1@git.server/repo_name@main"
+    secret_id = "secret:123213123123123123123"  # nosec
```
Comment (yanksyoon) [inline tests/unit/test_charm.py:49]: Nitpick:
Would secrets.token_hex(16) be applicable here?

**Review by yanksyoon — APPROVED** [review]
Minor suggestions, LGTM!

File: charmcraft.yaml (hunk 8 above)
Comment (arturo-seijas) [reply charmcraft.yaml:99]: No, that's why you can specify a private key to authenticate

**Review by mthaddon — APPROVED** [review]

**Review by yanksyoon — APPROVED** [review]

**Pull request merged**